experiment/telegram/telegram_otherwise.go

```diff
@@ -0,0 +1,184 @@
+// +build !cgo
+
+// Package telegram contains the Telegram network experiment. This file
+// in particular is a pure-Go implementation of that.
+//
+// See https://github.com/ooni/spec/blob/master/nettests/ts-020-telegram.md.
+package telegram
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/ooni/netx/handlers"
+	"github.com/ooni/netx/httpx"
+	"github.com/ooni/probe-engine/experiment"
+	"github.com/ooni/probe-engine/experiment/handler"
+	"github.com/ooni/probe-engine/log"
+	"github.com/ooni/probe-engine/model"
+	"github.com/ooni/probe-engine/session"
+)
+
+const (
+	testName    = "telegram"
+	testVersion = "0.0.2"
+)
+
+// Config contains the experiment config.
+type Config struct{}
+
+// TestKeys contains telegram test keys.
+type TestKeys struct {
+	TelegramHTTPBlocking bool   `json:"telegram_http_blocking"`
+	TelegramTCPBlocking  bool   `json:"telegram_tcp_blocking"`
+	TelegramWebFailure   string `json:"telegram_web_failure"`
+	TelegramWebStatus    string `json:"telegram_web_status"`
+}
+
+type measurer struct {
+	client *http.Client
+	logger log.Logger
+	tk     TestKeys
+}
+
+func (m *measurer) request(
+	ctx context.Context, method, scheme, address, port string,
+) (*http.Response, error) {
+	URL := url.URL{}
+	URL.Scheme = scheme
+	if port != "" {
+		URL.Host = net.JoinHostPort(address, port)
+	} else {
+		URL.Host = address
+	}
+	m.logger.Debugf("telegram: %s %s...", method, URL.String())
+	req, err := http.NewRequest(method, URL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	return m.client.Do(req)
+}
+
+func (m *measurer) measureDC(ctx context.Context) {
+	// "If all TCP connections on ports 80 and 443 to Telegram's access point
+	// IPs fail we consider Telegram to be blocked. [...]
+	//
+	// If at least an HTTP request [to an access point] returns back a
+	// response, we consider Telegram to not be blocked"
+	m.tk.TelegramTCPBlocking = true
+	m.tk.TelegramHTTPBlocking = true
+	var addresses = []string{
+		"149.154.175.50", "149.154.167.51", "149.154.175.100",
+		"149.154.167.91", "149.154.171.5",
+	}
+	for _, addr := range addresses {
+		for _, port := range []string{"80", "443"} {
+			const alwaysHTTP = "http" // note: it's intended to use HTTP with 443
+			resp, err := m.request(ctx, "POST", alwaysHTTP, addr, port)
```

Comment:
Is doing this in fact equivalent to just doing a TCP connection to the IP and port?

Reply:
I suspect that if the connection is aborted or reset after you send the HTTP POST payload you will still fall into the `if err != nil` below and result in marking the result as `tcp_blocking`, while in fact it's a different scenario.

Reply:
So, does `tcp_blocking` in this context mean only failure to connect and everything else is HTTP blocking?